man/lockfile.man
Please read the README file in this directory first.
.ex
.Id $Id$
.TH LOCKFILE 1 \*(Dt BuGless
.na
.SH NAME
lockfile \- conditional semaphore-file creator
.SH SYNOPSIS
.B lockfile
.I "\fB\-\fPsleeptime"
|
.I "\fB\-r \fPretries"
|
.if n .ti +0.5i
.I "\fB\-l \fPlocktimeout"
|
.I "\fB\-s \fPsuspend"
|
.B "\-!"
|
.B "\-ml"
|
.B "\-mu"
|
.I filename
\&.\|.\|.
.ad
.SH DESCRIPTION
.B lockfile
can be used to create one or more
.I semaphore
.IR files .
If lockfile can't create all the specified files (in the specified order),
it waits
.I sleeptime
(defaults to @DEFlocksleep@) seconds and retries the last file that didn't
succeed.  You can specify the number of
.I retries
to do until failure is returned.
If the number of
.I retries
is \-1 (default, i.e.,
.BR \-r\-1 )
lockfile will retry forever.
.PP
If the number of
.I retries
expires before all files have been created, lockfile returns failure and
removes all the files it created up till that point.
.PP
Using lockfile as the condition of a loop in a shell script can be done
easily by using the
.B \-!
flag to invert the exit status.  To prevent infinite loops, failures
for any reason other than the lockfile already existing are not
inverted to success but rather are still returned as failures.
.PP
All flags can be specified anywhere on the command line, they will be
processed when encountered.  The command line is simply parsed from
left to right.
.PP
All files created by lockfile will be read-only, and therefore
will have to be removed with
.B rm
.BR \-f .
.PP
If you specify a
.I locktimeout
then a lockfile will be removed by force after locktimeout seconds have
passed since the lockfile was last modified/created (most likely by some
other program that unexpectedly died a long time ago, and hence could not clean
up any leftover lockfiles).  Lockfile is clock skew immune.  After a lockfile
has been removed by force, a suspension of
.I suspend
seconds (defaults to @DEFsuspend@) is taken into account, in order to prevent
the inadvertent immediate removal of any newly created lockfile by another
program (compare
.BR SUSPEND
in
.BR procmail (1)).
.SS "Mailbox locks"
If the permissions on the system mail spool directory allow it, or if lockfile
is suitably setgid, it will be able to lock and unlock your system mailbox by
using the options
.B "\-ml"
and
.B "\-mu"
respectively.
.SH EXAMPLES
Suppose you want to make sure that access to the file "important" is
serialised, i.e., no more than one program or shell script should be allowed
to access it.  For simplicity's sake, let's suppose that it is a shell
script.  In this case you could solve it like this:
.RS
\&.\|.\|.
lockfile important.lock
\&.\|.\|.
access_"important"_to_your_hearts_content
\&.\|.\|.
rm \-f important.lock
\&.\|.\|.
.RE
Now if all the scripts that access "important" follow this guideline, you
will be assured that at most one script will be executing between the
`lockfile' and the `rm' commands.
.SH ENVIRONMENT
.TP 2.3i
.B LOGNAME
used as a hint to determine the invoker's loginname
.SH FILES
.TP 2.3i
.B /etc/passwd
to verify and/or correct the invoker's loginname (and to find out his HOME
directory, if needed)
.TP
.B @MAILSPOOLDIR@$LOGNAME@DEFlockext@
lockfile for the system mailbox, the environment variables present in here
will not be taken from the environment, but will be determined by looking
in /etc/passwd
.SH "SEE ALSO"
.na
.nh
.BR rm (1),
.BR mail (1),
.BR sendmail (8),
.BR procmail (1)
.hy
.ad
.SH DIAGNOSTICS
.TP 2.3i
Filename too long, .\|.\|.
Use shorter filenames.
.TP
Forced unlock denied on "x"
No write permission in the directory where lockfile "x" resides, or more than
one lockfile trying to force a lock at exactly the same time.
.TP
Forcing lock on "x"
Lockfile "x" is going to be removed by force because of a timeout
(compare
.BR LOCKTIMEOUT
in
.BR procmail (1)).
.TP
Out of memory, .\|.\|.
The system is out of swap space.
.TP
Signal received, .\|.\|.
Lockfile will remove anything it created till now and terminate.
.TP
Sorry, .\|.\|.
The
.I retries
limit has been reached.
.TP
Truncating "x" and retrying lock
"x" does not seem to be a valid filename.
.TP
Try praying, .\|.\|.
Missing subdirectories or insufficient privileges.
.SH BUGS
Definitely less than one.
.SH WARNINGS
The behavior of the
.B \-!
flag, while useful, is not necessarily intuitive or consistent.  When
testing lockfile's return value, shell script writers should consider
carefully whether they want to use the
.B \-!
flag, simply reverse the test, or do a switch on the exact exitcode.
In general, the
.B \-!
flag should only be used when lockfile is the conditional of a loop.
.SH MISCELLANEOUS
Lockfile is NFS-resistant and eight-bit clean.
.SH NOTES
Calling up lockfile with the \-@HELPOPT1@ or \-@HELPOPT2@ options will cause
it to display a command-line help page.  Calling it up with the \-@VERSIONOPT@
option will cause it to display its version information.
.PP
Multiple
.B \-!
flags will toggle the return status.
.PP
Since flags can occur anywhere on the command line, any filename starting
with a '-' has to be preceded by './'.
.PP
The number of
.I retries
will not be reset when any following file is being created (i.e., they are
simply used up).  It can, however, be reset by specifying
.RI \-r newretries
after every file on the command line.
.PP
Although files with any name can be used as lockfiles, it is common practice
to use the extension `.lock' to lock mailfolders (it is appended to the
mailfolder name).  In case one does not want to have to worry about too long
filenames and does not have to conform to any other lockfilename convention,
then an excellent way to generate a lockfilename corresponding to some already
existing file is by taking the prefix `lock.' and appending the i-node number
of the file which is to be locked.


HISTORY
	Only the last entry is complete, the others might have been condensed.

1990/12/07: v1.00
1990/12/12: v1.01
1991/02/04: v1.02
1991/02/13: v1.10
1991/02/21: v1.20
1991/02/22: v1.21
1991/03/01: v1.30
1991/03/15: v1.35
	    Started using RCS to manage the source
1991/06/04: v1.99
1991/06/10: v2.00
1991/06/11: v2.01
1991/06/12: v2.02
1991/06/20: v2.03
1991/07/04: v2.10
1991/07/12: v2.11
1991/10/02: v2.20 (never released)
1991/10/18: v2.30
	    Reached the doubtful milestone of having a source file (regexp.c)
	       which provokes a compiler error on an old compiler
	       (if using the optimiser)
1991/10/22: v2.31
1991/12/05: v2.40
1991/12/13: v2.50
1992/01/22: v2.60
1992/01/31: v2.61
1992/04/30: v2.70
1992/07/01: v2.71
	    Gave procmail, formail, lockfile and mailstat a more verbose
	       command line help (called up by -h or -?)
1993/02/04: v2.80
	    Started using CVS to manage the source (god's gift to programmers)
	    Changes to the installation scripts:
	       - the autoconf script now performs a reliability test on kernel
		 locking support
	       - reached the doubtful milestone of consistently crashing the
		 kernel on a Convex by running the locktst program
1993/02/19: v2.81
1993/06/02: v2.82 (never really released, was only available as prerelease 4)
	    Worked my way around the !@#$%^&*() POSIX setgid() semantics (if
	       your OS supports setrgid() or setregid())
1993/07/01: v2.90
	    Condition lines in recipes can now be started with a leading `*',
	       there is no longer a need to count condition lines, simply
	       set the number to zero, and let procmail find out by itself
1993/07/02: v2.91
	    Reached the doubtful milestone to sometimes crash an Ultrix
	       machine (due to the lockingtests, not procmail itself)
1994/06/14: v3.00
	    Changes to procmail:
	       - Changed the semantics of the TRAP keyword.  In order to
		 make procmail accept the exitcode it returns, you now have
		 to set EXITCODE=""
	       - It was still occasionally trying to lock /dev/null, which
		 is of course silly, fixed that
	       - Taught it about `nesting recipes'; they allow parts of
		 an rcfile to be grouped hierarchically
	       - Fixed a discrepancy with /bin/sh backquote expansion in
		 environment assignments (preserving all spaces)
	       - Logs its pid and a timestamp when VERBOSE=on
	       - Caused the regular TIMEOUT to break a `hanging' kernel lock
	       - SIGUSR1 and SIGUSR2 can be used to turn on and off verbose
		 logging
	       - Worked around a bug in the `ANSI'-compiler of Domain/OS
	       - Procmail and lockfile now inherit any ignore status of most
		 regular signals (fixes a problem with some buggy shells)
	       - Optionally reads in a global rcfile (/etc/procmailrc)
		 before doing regular delivery (which includes the new
		 keyword: DROPPRIVS)
	       - Can pipe the mail to stdout on request
	       - Moved the "Reiterating kernel lock" diagnostic into the
		 "extended" (i.e. VERBOSE=on) section
	       - Tightened the loop when skipping comments in rcfiles (for
		 a slight speedup)
	       - Added support for filesystems not capable of creating
		 hardlinks
	       - Tightened the security check on initial absolute rcfiles
		 (they sometimes can't be world writable)
	       - Weighted scoring on conditions
	       - Ability to inline parse ${var-text} and ${var:-text}
	       - Ability to inline parse ${var+text} and ${var:+text}
	       - Skipping spaces after "!" and "$" on condition lines
	       - Implicit delivery somehow got broken: fixed
	       - Default umask is always 077 now for deliverymode
	       - Extended ^FROM_DAEMON and ^FROM_MAILER macro regexps again
	       - The -f option became less strict, everyone can use it now,
		 except that unpriviliged users will get an additional >From_
		 they didn't bargain for (in order to make fakes identifiable)
	       - The date on the From_ line can now be refreshed with -f-
	       - Introduced new recipe flags: E and e (else and error)
	       - Nested blocks clone procmail on a 'c' flag
	       - Introduced the EXITCODE special variable
	       - Implicit delivery mode is now entered if argv[0] doesn't start
		 with the word `procmail'
	       - Fixed the BSD support for kernel-locking only operation
	       - Taught the regexp engine about \< and \>
	       - Fixed bug present on some systems; caused the body to be
		 munged when filtering headers only
	       - Added -o option (makes procmail override the From_ lines, like
		 it used to)
	       - -p and -m together shrink the set of preset variables to the
		 bare minimum
	       - -p is not supported alongside -d anymore
	       - /etc/procmailrcs/ is the place for optional privileged
		 rcfiles in -m mailfilter mode
	       - Switched the meanings of SIGUSR1 and SIGUSR2
	       - The 'a' flag didn't work correctly after filter recipes
	       - Changed the permissions on the lockfile, writing zero in it
	       - Check the permissions on the existing system mailbox, correct
		 them if necessary
	       - Clean up zombies more often
	    Changes to formail:
	       - Fixed a sender-determination-weight problem, it mixed up
		 the weights when autoreplying and when regenerating the From_
		 line (and thus didn't always pick the optimal field)
	       - Pays attention to the exitcode of the programs it started
	       - Accepts simultaneous -X and -k options
	       - Fixed a bug introduced in v2.82 in formail when using
		 the -x and the -k options simultaneously
	       - Rearranged the weights for "-rt" (made From: more important)
	       - Parsed return-addresses starting with a \ incorrectly
		 (causing it to coredump on occasion)
	       - Supports the -s option withouth a program argument
	       - Recognise extra UUCP >From_ lines
	       - Introduced the -B option to split up BABYL rmail files
	       - It regards and generates a FILENO variable (for easy
		 numbering)

src/lockfile.c
/************************************************************************
 *	lockfile - The conditional semaphore-file creator		*
 *									*
 *	It has been designed to be able to be run sgid mail or		*
 *	any gid you see fit (in case your mail spool area is *not*	*
 *	world writable, but group writable), without creating		*
 *	security holes.							*
 *									*
 *	Seems to be relatively bug free.				*
 *									*
 *	Copyright (c) 1990-1999, S.R. van den Berg, The Netherlands	*
 *	Copyright (c) 1999-2001, Philip Guenther, The United States	*
 *							of America	*
 *	#include "../README"						*
 ************************************************************************/
#ifdef RCS
static /*const*/char rcsid[]=
 "$Id$";
#endif
static /*const*/char rcsdate[]="$Date$";
#include "includes.h"
#include "sublib.h"
#include "exopen.h"
#include "mcommon.h"
#include "authenticate.h"
#include "lastdirsep.h"
#include "../patchlevel.h"

static volatile int exitflag;
pid_t thepid;
uid_t uid;
gid_t sgid;
const char dirsep[]=DIRSEP;
static const char lockext[]=DEFlockext,
 nameprefix[]="lockfile: ",lgname[]="LOGNAME";

static void failure P((void))				      /* signal trap */
{ exitflag=2;					       /* merely sets a flag */
}
				    /* see locking.c for comment on xcreat() */
static int xcreat(name,tim)const char*const name;time_t*const tim;
{ char*p;int j= -1;size_t i;struct stat stbuf;
  i=lastdirsep(name)-name;
  if(!(p=malloc(i+UNIQnamelen)))
     return exitflag=1;
  memcpy(p,name,i);
  if(unique(p,p+i,i+UNIQnamelen,LOCKperm,0,doCHECK|doLOCK))
     stat(p,&stbuf),*tim=stbuf.st_mtime,j=myrename(p,name);
  free(p);
  return j;
}

void elog(a)const char*const a;
{ write(STDERR,a,strlen(a));
}

void nlog(a)const char*const a;
{ elog(nameprefix);elog(a);  /* decent error messages should start with this */
}

static PROGID;

int main(argc,argv)int argc;const char*const argv[];
{ const char*const*p;char*cp;uid_t uid;
  int sleepsec,retries,invert,force,suspend,retval=EXIT_SUCCESS,virgin=1;
  static const char usage[]="Usage: lockfile -v | -nnn | -r nnn | -l nnn \
| -s nnn | -! | -ml | -mu | file ...\n";
  if(argc<=1)			       /* sanity check, any argument at all? */
     goto usg;
  sleepsec=DEFlocksleep;retries= -1;suspend=DEFsuspend;thepid=getpid();force=0;
  uid=getuid();signal(SIGPIPE,SIG_IGN);
  if(setuid(uid)||geteuid()!=uid)		  /* resist setuid operation */
sp:{ nlog("Unable to give up special permissions");
     return EX_OSERR;
   }
again:
  invert=(char*)progid-(char*)progid;qsignal(SIGHUP,failure);
  qsignal(SIGINT,failure);qsignal(SIGQUIT,failure);qsignal(SIGTERM,failure);
  for(p=argv;--argc>0;)
     if(*(cp=(char*)*++p)=='-')
	for(cp++;;)
	 { char*cp2=cp;int i;
	   switch(*cp++)
	    { case '!':invert^=1;		      /* invert the exitcode */
		 continue;
	      case 'r':case 'l':case 's':
		 if(!*cp&&(cp=(char*)*++p,!--argc)) /* concatenated/seperate */
		  { p--;
		    goto eusg;
		  }
		 i=strtol(cp,&cp,10);
		 switch(*cp2)
		  { case 'r':retries=i;
		       goto checkrdec;
		    case 'l':force=i;
		       goto checkrdec;
		    default:
		       if(i<0)			    /* suspend should be >=0 */
			  goto eusg;
		       suspend=i;
		       goto checkrdec;
		  }
	      case VERSIONOPT:elog("lockfile");elog(VERSION);
		    elog("\nYour system mailbox's lockfile:\t");
		    elog(auth_mailboxname(auth_finduid(getuid(),0)));
		    elog(lockext);elog("\n");
		  goto xusg;
	      case HELPOPT1:case HELPOPT2:elog(usage);
		 elog(
 "\t-v\tdisplay the version number and exit\
\n\t-nnn\twait nnn seconds between locking attempts\
\n\t-r nnn\tmake at most nnn retries before giving up on a lock\
\n\t-l nnn\tset locktimeout to nnn seconds\
\n\t-s nnn\tsuspend nnn seconds after a locktimeout occurred\
\n\t-!\tinvert the exitcode of lockfile\
\n\t-ml\tlock your system mail-spool file\
\n\t-mu\tunlock your system mail-spool file\n");
		 goto xusg;
	      default:
		 if(sleepsec>=0)	    /* is this still the first pass? */
		  { if((sleepsec=strtol(cp2,&cp,10))<0)
		       goto eusg;
checkrdec:	    if(cp2==cp)
eusg:		     { elog(usage);		    /* regular usage message */
xusg:		       retval=EX_USAGE;
		       goto nfailure;
		     }
		  }
		 else		      /* no second pass, so leave sleepsec<0 */
		    strtol(cp2,&cp,10);		   /* and discard the number */
		 continue;
	      case 'm':		  /* take $LOGNAME as a hint, check if valid */
	       { auth_identity*pass;static char*ma;
		 if(*cp&&cp[1]||ma&&sleepsec>=0)	     /* second pass? */
		    goto eusg;
		 if(!ma)			/* ma initialised last time? */
		  { if(!((ma=(char*)getenv(lgname))&&
		      (pass=auth_finduser(ma,0))&&
		      auth_whatuid(pass)==uid||
		     (pass=auth_finduid(uid,0))))
		     { nlog("Can't determine your mailbox, who are you?\n");
		       goto nfailure;	 /* panic, you're not in /etc/passwd */
		     }
		    ;{ const char*p;
		       if(!*(p=auth_mailboxname(pass))||
			!(ma=malloc(strlen(p)+STRLEN(lockext)+1)))
			  goto outofmem;
		       strcat(strcpy(ma,p),lockext);
		     }
		  }
		 switch(*cp)
		  { default:
		       goto eusg;		    /* somebody goofed again */
		    case 'l':				 /* lock the mailbox */
		       if(sleepsec>=0)			      /* first pass? */
			{ cp=ma;
			  goto stilv;			    /* yes, lock it! */
			}
		    case 'u':			       /* unlock the mailbox */
		       if(unlink(ma))
			{ nlog("Can't unlock \"");elog(ma);elog("\"");
			  if(*cp=='l')	 /* they messed up, give them a hint */
			     elog(" again,\n already dropped my privileges");
			  elog("\n");
			}
		       else
			  virgin=0;
		  }

	       }
	      case '\0':;
	    }
	   break;
	 }
     else if(sleepsec<0)      /* second pass, release everything we acquired */
	unlink(cp);
     else
      { time_t t;int permanent;gid_t gid=getgid();
	if(setgid(gid)||getegid()!=gid)	      /* just to be on the safe side */
	   goto sp;
stilv:	virgin=0;permanent=nfsTRY;
	while(0>xcreat(cp,&t))				     /* try and lock */
	 { struct stat stbuf;
	   if(exitflag)					    /* time to stop? */
	    { if(exitflag==1)		     /* was it failure() or malloc() */
outofmem:	 retval=EX_OSERR,nlog("Out of memory");
	      else
		 retval=EX_TEMPFAIL,nlog("Signal received");
	      goto lfailure;
	    }
	   switch(errno)		    /* why did the lock not succeed? */
	    { case EEXIST:			  /* hmmm..., by force then? */
		 if(force&&!lstat(cp,&stbuf)&&force<t-stbuf.st_mtime)
		  { nlog(unlink(cp)?"Forced unlock denied on \"":
		     "Forcing lock on \"");
		    elog(cp);elog("\"\n");sleep(suspend);	/* important */
		  }
		 else					   /* no forcing now */
		    switch(retries)    /* await your turn like everyone else */
		     { case 0:nlog("Sorry");retval=EX_CANTCREAT;
			  goto lfailure;      /* patience exhausted, give up */
		       default:retries--;		      /* count sheep */
		       case -1:sleep(sleepsec);		     /* wait and see */
		     }
		 break;
	      case ENOSPC:
#ifdef EDQUOT
	      case EDQUOT:
#endif
	      case ENOENT:case ENOTDIR:case EIO:/*case EACCES:*/
		 if(!--permanent)	 /* NFS sporadically generates these */
		  { sleep(sleepsec);	/* unwarranted, so ignore them first */
		    continue;
		  }
	      default:		     /* but, it seems to persist, so give up */
		 nlog("Try praying");retval=EX_UNAVAILABLE;
#ifdef ENAMETOOLONG
		 goto lfailure;
	      case ENAMETOOLONG:
		 if(0<(permanent=strlen(cp)-1)&&      /* can we truncate it? */
		  !strchr(dirsep,cp[permanent-1]))
		  { nlog("Truncating \"");elog(cp);	      /* then try it */
		    elog("\" and retrying lock\n");cp[permanent]='\0';
		    break;
		  }				     /* otherwise, forget it */
		 nlog("Filename too long");retval=EX_UNAVAILABLE;
#endif
lfailure:	 elog(", giving up on \"");elog(cp);elog("\"\n");
nfailure:	 sleepsec= -1;argc=p-argv;		    /* mark sleepsec */
		 goto again;
	    }  /* for second pass, and adjust argc to the no. of args parsed */
	   permanent=nfsTRY;	       /* refresh the NFS-error-ignore count */
	 }
      }
  if(retval==EXIT_SUCCESS&&virgin)	 /* any errors?	 did we do anything? */
usg:
   { elog(usage);
     return EX_USAGE;
   }
  if(invert)
     switch(retval)			 /* we only invert the regular cases */
      { case EXIT_SUCCESS:
	   return EX_CANTCREAT;
	case EX_CANTCREAT:
	   return EXIT_SUCCESS;
      }
  return retval;			       /* all other exitcodes remain */
}

void*tmalloc(len)const size_t len;				     /* stub */
{ void*p;
  if(!(p=malloc(len)))
     exitflag=1;				     /* signal out of memory */
  return p;
}

void tfree(p)void*const p;					     /* stub */
{ free(p);
}

int ropen(name,mode,mask)const char*const name;const int mode;
 const mode_t mask;
{ return open(name,mode,mask);					     /* stub */
}

int rwrite(fd,a,len)const int fd;const void*const a;const int len;   /* stub */
{ return write(fd,a,len);
}

int rclose(fd)const int fd;					     /* stub */
{ return close(fd);
}

void writeerr(a)const char*const a;				     /* stub */
{
}

char*cstr(a,b)char*const a;const char*const b;			     /* stub */
{ return 0;
}

void ssleep(seconds)const unsigned seconds;			     /* stub */
{ sleep(seconds);
}
